The ContentFilter method continueAfterWillSendRequest should use a completion handler

pvollan opened the pull request:
#### f5665ae82cc3854a847d27d097f30ab01b092137
<pre>
The ContentFilter method continueAfterWillSendRequest should use a completion handler
<a href="https://bugs.webkit.org/show_bug.cgi?id=301189">https://bugs.webkit.org/show_bug.cgi?id=301189</a>
<a href="https://rdar.apple.com/151461666">rdar://151461666</a>

Reviewed by Chris Dumez and Sihui Liu.

The ContentFilter method continueAfterWillSendRequest should use a completion handler to avoid blocking
the main thread waiting for a decision that can take a significant amount of time. In particular, it is
the NetworkExtensionContentFilter that can take a long time. For this reason, it is currently doing the
work on a non-main thread, and we have a binary semaphore that we use to wait synchronously for the
decision. Performing this work on the main thread can block other network requests. To avoid blocking
the main thread waiting for a semaphore, we should instead use a completion handler. This patch adds no
new threads, but is instead introducing a completion handler that is called when the decision is ready.

Before this change, the responsiveness timer in the UI process would terminate the Networking process
if the content filter was blocking the main thread for more than 6s. This issue is resolved with this
patch.

Test: contentfiltering/delay-willsendrequest-decision.html

* Source/WebCore/loader/ContentFilter.cpp:
(WebCore::ContentFilter::continueAfterWillSendRequest):
(WebCore::ContentFilter::ContentFilterCallbackAggregator::~ContentFilterCallbackAggregator):
(WebCore::ContentFilter::ContentFilterCallbackAggregator::didReceivePlatformContentFilterDecision):
(WebCore::ContentFilter::ContentFilterCallbackAggregator::ContentFilterCallbackAggregator):
(WebCore::ContentFilter::anyContentFilterIsEnabled const):
* Source/WebCore/loader/ContentFilter.h:
(WebCore::ContentFilter::ContentFilterCallbackAggregator::create):
* Source/WebCore/platform/PlatformContentFilter.h:
(WebCore::PlatformContentFilter::isEnabled const):
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.h:
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
(WebCore::NetworkExtensionContentFilter::willSendRequest):
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
(WebCore::ParentalControlsContentFilter::willSendRequest):
* Source/WebCore/testing/MockContentFilter.cpp:
(WebCore::MockContentFilter::willSendRequest):
* Source/WebCore/testing/MockContentFilter.h:
* Source/WebCore/testing/MockContentFilterSettings.cpp:
(WebCore::MockContentFilterSettings::setWillSendRequestDecisionDelay):
* Source/WebCore/testing/MockContentFilterSettings.h:
(WebCore::MockContentFilterSettings::willSendRequestDecisionDelay const):
* Source/WebCore/testing/MockContentFilterSettings.idl:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startContentFiltering):
(WebKit::NetworkResourceLoader::startWithServiceWorker):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/302627@main">https://commits.webkit.org/302627@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e343d2ec4cf7633085f2caafce15c099747e9043

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129698 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1959 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40554 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137086 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81159 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/28a7f8d4-ba8c-4a96-9afb-e08ee6fc9d3a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131569 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1907 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1848 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98800 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66628 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a3223afe-69e9-4efe-8314-72009727b143) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132645 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1460 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116160 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79475 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/eb4f4bbb-861d-47f4-b7fd-86747adf59c1) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1378 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34293 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80359 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109855 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34794 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139568 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1753 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1643 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107310 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1798 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112507 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107178 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1416 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31007 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54503 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20238 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1826 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65195 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1640 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1675 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1747 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -86,12 +86,13 @@ std::unique_ptr<ContentFilter> ContentFilter::create(ContentFilterClient& client
         return nullptr;
 
-    return makeUnique<ContentFilter>(WTFMove(filters), client);
+    return new ContentFilter(WTFMove(filters), client);
```
aren't you missing `adoptRef()` here? Seems like this would be leaking otherwise.

pvollan replied:
Good catch! I will address this issue and the other comments in follow-up PRs.

Thanks for reviewing!

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +124,43 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
 
+void ContentFilter::continueAfterWillSendRequest(ResourceRequest&& request, const ResourceResponse& redirectResponse, CompletionHandler<void(ResourceRequest)>&& completionHandler)
+{
+    Ref protectedClient { m_client.get() };
```
We don't do local protector variables in new code.

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +124,43 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+        return;
+    }
+
+    m_workQueue->dispatch([protectedThis = RefPtr { this }, completionHandler = WTFMove(completionHandler), request = crossThreadCopy(WTFMove(request)), redirectResponse = WTFMove(redirectResponse), protectedClient = WTFMove(protectedClient)] mutable {
```
`= Ref { *this }`? Why would you crossThreadCopy the request but not the redirect response. Looks like a thread safety bug.

pvollan replied:
The work queue has been removed in the latest patch.

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +124,43 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+            if (blocked)
+                request = ResourceRequest();
+#if !LOG_DISABLED
+            if (request != originalRequest)
```
Do we really need full requests here and `operator!=`? Or do we only need to check if the URL changed? If so, we could merely store URLs.

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +124,43 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+                request = ResourceRequest();
+#if !LOG_DISABLED
+            if (request != originalRequest)
+                LOG(ContentFiltering, "ContentFilter changed request url to <%{sensitive}s>.\n", originalRequest.url().string().ascii().data());
```
Why `ascii()` and not `utf8()` (which I think is more efficient).

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -391,6 +429,15 @@ bool ContentFilter::isWebContentRestrictionsUnblockURL(const URL& url)
 
+bool ContentFilter::anyContentFilterIsEnabled() const
+{
+    for (Ref contentFilter : m_contentFilters) {
```
Looks like a good use case for `std::ranges::any_of()`

cdumez commented on Source/WebCore/testing/MockContentFilter.cpp:
```diff
@@ -73,6 +73,11 @@ void MockContentFilter::willSendRequest(ResourceRequest& request, const Resource
 
+    if (MockContentFilterSettings::singleton().delayWillSendRequestDecision()) {
+        LOG(ContentFiltering, "MockContentFilter delaying decision in willSendRequest\n");
+        WTF::sleep(10_s);
```
What's this? Delaying things by 10 seconds in our testing is a very long time.

pvollan replied:
That is a good point! This delay is needed to reproduce the issue, since it will block the main thread in the Networking process by 10s (before this patch). By having a delay longer than the Networking process responsiveness timer (6s), we can reproduce that the UI process is terminating the blocked Networking process, which is causing this issue.

cdumez commented on Source/WebCore/testing/MockContentFilterSettings.h:
```diff
@@ -75,9 +75,13 @@ class MockContentFilterSettings {
     const String& modifiedRequestURL() const { return m_modifiedRequestURL; }
     WEBCORE_TESTSUPPORT_EXPORT void setModifiedRequestURL(const String&);
 
+    bool delayWillSendRequestDecision() const { return m_delayWillSendRequestDecision; }
```
Missing boolean prefix.

cdumez commented on Source/WebCore/testing/MockContentFilterSettings.h:
```diff
@@ -87,6 +91,7 @@ class MockContentFilterSettings {
     Decision m_unblockRequestDecision { Decision::Block };
     String m_blockedString;
     String m_modifiedRequestURL;
+    bool m_delayWillSendRequestDecision { false };
```
Missing boolean prefix.

cdumez commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -2052,23 +2061,29 @@ void NetworkResourceLoader::startWithServiceWorker()
 {
     LOADER_RELEASE_LOG("startWithServiceWorker:");
 
+    CompletionHandler<void(ResourceRequest)> completionHandler = [protectedThis = RefPtr { this }](ResourceRequest request) {
```
`= Ref { *this }`

cdumez commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
```diff
@@ -76,7 +76,7 @@ class Entry;
 
 class NetworkResourceLoader final
-    : public RefCounted<NetworkResourceLoader>
+    : public ThreadSafeRefCounted<NetworkResourceLoader>
```
This class has data members like Strings that are not thread safe. How do we make sure they only get used on the right things (including destruction).

pvollan replied:
With the new refactoring, it is no longer required to make NetworkResourceLoader ThreadSafeRefCounted.

cdumez commented on Source/WebCore/loader/ContentFilter.h:
```diff
@@ -44,14 +45,13 @@ class ResourceRequest;
 class SubstituteData;
 
-class ContentFilter : public CanMakeWeakPtr<ContentFilter>, public CanMakeCheckedPtr<ContentFilter> {
+class ContentFilter : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<ContentFilter> {
```
This class has non-thread safe data members (like URLs). How do we make sure they're only used from the right thread (including destruction)?

szewai replied:
I think we also need to make `PlatformContentFilter` (stored in `Container m_contentFilters`) thread-safe first as this patch invokes its functions on different threads.

pvollan replied:
With the new refactoring, it is no longer required to make ContentFilter ThreadSafeRefCounted.

Thanks for reviewing, all!

cdumez commented on Source/WebCore/loader/ContentFilter.h:
```diff
@@ -121,6 +126,9 @@ class ContentFilter : public CanMakeWeakPtr<ContentFilter>, public CanMakeChecke
     bool m_responseReceived { false };
+    unsigned m_platformCompletionHandlersCalled { 0 };
+    unsigned m_platformFiltersAllowed { 0 };
+    std::optional<CompletionHandler<void(ResourceRequest)>> m_willSendRequestCompletionHandler;
```
Why is this using `std::optional`? WTF::CompletionHandler already has a null state, isn't it enough, do we need to distinguish std::nullopt vs the CompletionHandler being null somehow?

pvollan replied:
Good catch! This member has been removed in the latest patch.

cdumez commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -2052,23 +2061,32 @@ void NetworkResourceLoader::startWithServiceWorker()
 {
     LOADER_RELEASE_LOG("startWithServiceWorker:");
 
+    CompletionHandler<void(ResourceRequest)> completionHandler = [protectedThis = Ref { *this }](ResourceRequest request) {
```
`ResourceRequest&&` or `const ResourceRequest&` but we should not pass by value.

pvollan replied:
Ah, good catch, fixed in latest patch!

cdumez commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -284,21 +284,30 @@ void NetworkResourceLoader::startRequest(const ResourceRequest& newRequest)
-    }
-    return true;
+
+    CompletionHandler<void(ResourceRequest)> completion = [contentFilter = RefPtr { m_contentFilter }, completionHandler = WTFMove(completionHandler)](ResourceRequest request) mutable {
```
We should not pass `ResourceRequest` by value.

cdumez replied:
Why can't we just capture the `contentFilter` local variable here?

cdumez commented on Source/WebCore/testing/MockContentFilterSettings.h:
```diff
@@ -87,6 +91,7 @@ class MockContentFilterSettings {
     Decision m_unblockRequestDecision { Decision::Block };
     String m_blockedString;
     String m_modifiedRequestURL;
+    double m_willSendRequestDecisionDelay { 0 };
```
This should probably be using `Seconds` type. `double` is so ambiguous for this.

cdumez commented on Source/WebCore/testing/MockContentFilterSettings.cpp:
```diff
@@ -92,6 +92,12 @@ const String& MockContentFilterSettings::unblockRequestURL() const
 
+void MockContentFilterSettings::setWillSendRequestDecisionDelay(double willSendRequestDecisionDelay)
+{
+    m_willSendRequestDecisionDelay= willSendRequestDecisionDelay;
```
Missing space before `=`

cdumez commented on Source/WebCore/testing/MockContentFilter.cpp:
```diff
@@ -94,6 +95,46 @@ void MockContentFilter::willSendRequest(ResourceRequest& request, const Resource
+
+    URL modifiedRequestURL { request.url(), modifiedRequestURLString };
+    if (!modifiedRequestURL.isValid()) {
+        LOG(ContentFiltering, "MockContentFilter failed to convert %s to a  URL.\n", modifiedRequestURL.string().ascii().data());
```
Why is this using `ascii()`?

pvollan replied:
Good point! This will be fixed in the upcoming patch.

Thanks for reviewing!

cdumez commented on Source/WebCore/testing/MockContentFilter.cpp:
```diff
@@ -94,6 +95,46 @@ void MockContentFilter::willSendRequest(ResourceRequest& request, const Resource
+{
+    if (!enabled()) {
+        m_state = State::Allowed;
+        completionHandler(WTFMove(request), m_state);
```
We may want to use `CompletionHandlerCallingScope` in this function so that it is not so easy to forget calling the completion handler. It seems all the calls are identical anyway.

pvollan replied:
Ah, good point! Looking closer, it appears that `CompletionHandlerCallingScope` only accepts a completion handler with the type `CompletionHandler<void()>`, so it may not be applicable in this case.

cdumez replied:
As far as I can tell, it applies fine? You pass a CompletionHandler to `CompletionHandlerCallingScope` which is a `CompletionHandler<void()>` and have it capture your own completion handler and have the lambda call your own completion handler.

pvollan replied:
Yes, that is a good point! Fixed in the latest patch.

Thanks for reviewing!

cdumez commented on Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
```diff
@@ -114,6 +117,55 @@
+        }
+    }
+
+    RetainPtr protectedNSURLRequest = request.protectedNSURLRequest(DoNotUpdateHTTPBody);
```
We don't do variables with `protected` in the name anymore. maybe call this `nsRequest`?

cdumez commented on Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
```diff
@@ -114,6 +117,55 @@
+    auto blockPtr = makeBlockPtr([protectedThis = Ref { *this }, request = crossThreadCopy(WTFMove(request)), completionHandler = WTFMove(completionHandler)](NEFilterSourceStatus status, NSDictionary *decisionInfo) mutable {
+        RetainPtr<NSString> modifiedRequestURLString;
+        if (decisionInfo)
+            modifiedRequestURLString = decisionInfo[NEFilterSourceOptionsRedirectURL];
```
We probably want to use `checked_objc_cast<NSString>()` here instead of the assertion below.

cdumez commented on Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
```diff
@@ -114,6 +117,55 @@
+
+            URL modifiedRequestURL { modifiedRequestURLString.get() };
+            if (!modifiedRequestURL.isValid()) {
+                LOG(ContentFiltering, "NetworkExtensionContentFilter failed to convert modified URL string %@ to a  URL.\n", modifiedRequestURLString.get());
```
Too many spaces before `URL`

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,64 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+    ASSERT(m_state == State::Stopped || m_state == State::Filtering);
+
+    m_platformCompletionHandlersCalled = 0;
+    m_platformFiltersAllowed = 0;
```
I did not understand those 2 counters here yet, I'll look more after lunch but my initial gut feeling is that it looks fragile.

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,64 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+    m_willSendRequestCompletionHandler = WTFMove(completionHandler);
+
+    for (Ref platformContentFilter : m_contentFilters) {
+
```
Why the blank line here?

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,64 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+        URL originalURL = request.url();
+        CompletionHandler<void(ResourceRequest, PlatformContentFilter::State)> completion = [platformContentFilter, protectedThis = Ref { *this }, originalURL = WTFMove(originalURL)] (ResourceRequest request, PlatformContentFilter::State state) mutable {
+            ASSERT(RunLoop::isMain());
+            ++protectedThis->m_platformCompletionHandlersCalled;
```
Is there a way to leverage `WTF::CallbackAggregator` instead of dealing with counts? Seems it would be more robust.

pvollan replied:
That is a very good point. I am looking into this now.

Thanks for reviewing!

pvollan replied:
The standard `WTF::CallbackAggregator` didn't quite match in this case, since it only allows one type of completion handler, so I created a custom aggregator instead.

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,86 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+    LOG(ContentFiltering, "ContentFilter received request for <%{sensitive}s> with redirect response from <%{sensitive}s>.\n", request.url().string().ascii().data(), redirectResponse.url().string().ascii().data());
+    ASSERT(m_state == State::Stopped || m_state == State::Filtering);
+
+    class ContentFilterCallbackAggregator : public ThreadSafeRefCounted<ContentFilterCallbackAggregator> {
```
Can we move this out of the function maybe? This function is huge enough.

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,86 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+        return;
+    }
+
+    LOG(ContentFiltering, "ContentFilter received request for <%{sensitive}s> with redirect response from <%{sensitive}s>.\n", request.url().string().ascii().data(), redirectResponse.url().string().ascii().data());
```
Why is this still using `ascii()`?

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,86 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+
+        void didReceivePlatformContentFilterDecision(PlatformContentFilter& platformContentFilter, ResourceRequest&& request, PlatformContentFilter::State state)
+        {
+            m_platformCompletionHandlersCalled++;
```
`++m_platformCompletionHandlersCalled;`

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,86 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+            }
+
+            // Call completion handler when all platform completion handlers have been called.
+            if (m_platformCompletionHandlersCalled == m_contentFilter->m_contentFilters.size()) {
```
Normally, this is the kind of work that happens in the destructor of a CallbackAggregator, so that we don't have to use a counter. When all the completion handlers have been called, and given that they capture the callback aggregator, the destructor gets called and some logic occurs (the completion handle we usually pass to the callback aggregator constructor).

pvollan replied:
That is a good point! Fixed in the latest patch.

beidson commented on Source/WebCore/loader/ContentFilter.h:
```diff
@@ -44,14 +47,13 @@ class ResourceRequest;
 class SubstituteData;
 
-class ContentFilter : public CanMakeWeakPtr<ContentFilter>, public CanMakeCheckedPtr<ContentFilter> {
+class ContentFilter : public RefCountedAndCanMakeWeakPtr<ContentFilter> {
```
Reading the most recent version of the patch, I think the lifetime of the `ContentFilter` itself - as captured by lambdas - is correct.

But because code like this is tricky, and because someone might come along later and break assumptions without realizing it, I'd like to see this derive from `ThreadSafeRefCounted`, and then see a `RELEASE_ASSERT(isMainThread)` in `ContentFilter::~ContentFilter()`

pvollan replied:
That's a good point! In the latest patch, I went back to the original `CanMakeWeakPtr`/`CanMakeCheckedPtr`, since there wasn't a strict requirement for it to be reference counted.

beidson commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,73 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+    if (!anyContentFilterIsEnabled()) {
+        continueAfterWillSendRequest(request, redirectResponse);
+        completionHandler(WTFMove(request));
+        return;
```
I was originally confused as to why the synchronous version of `continueAfterWillSendRequest` remained, and I think it's just for this.

I think we should rename it to something like `continueAfterWillSendRequestInternal` and move it to private, so outsides (clients and specific platform filters) can't "accidentally" use it anymore.

pvollan replied:
That's a good point! The synchronous version is actually still in use in WK1. I didn't change the WK1 behavior, in order to minimize the changes in this patch. Would you be OK with potentially doing this in a follow-up patch?

Thanks for reviewing!

beidson commented on Source/WebCore/platform/PlatformContentFilter.h:
```diff
@@ -67,7 +67,9 @@ class PlatformContentFilter : public ThreadSafeRefCountedAndCanMakeThreadSafeWea
 
     virtual ~PlatformContentFilter() = default;
+    virtual bool isEnabled() const { return true; }
     virtual void willSendRequest(ResourceRequest&, const ResourceResponse&) = 0;
```
Similarly, can we get rid of these one? Moving all users to the new one?

pvollan replied:
This is still being called in WK1, so I didn't remove in order to minimize changes. Do you think it's acceptable to postpone this change to a possible follow-up patch?

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,73 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+ContentFilter::ContentFilterCallbackAggregator::~ContentFilterCallbackAggregator()
+{
+    ASSERT(m_callback);
+    CheckedPtr contentFilter = m_contentFilter;
```
We should reduce the scope of this variable to not cover the `m_callback` calls. It is too risky otherwise given that this is a CheckedPtr.

pvollan replied:
Good catch! Fixed in latest patch.

Thanks for reviewing!

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,73 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+
+void ContentFilter::ContentFilterCallbackAggregator::didReceivePlatformContentFilterDecision(PlatformContentFilter& platformContentFilter, ResourceRequest&& request, PlatformContentFilter::State state)
+{
+    CheckedPtr contentFilter = m_contentFilter;
```
We should always use the smallest scope possible for variables and this is especially true for CheckedPtr variable. We only need this variable in the `if (blocked && !m_blocked)` block in seems so let's move it there.

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,73 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
 
+ContentFilter::ContentFilterCallbackAggregator::~ContentFilterCallbackAggregator()
+{
+    ASSERT(m_callback);
```
Can we add assertions to distinguish what runs on the main thread vs what doesn't?

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,73 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+    if (state == PlatformContentFilter::State::Allowed)
+        ++m_platformFiltersAllowed;
+
+    bool blocked = state == PlatformContentFilter::State::Blocked;
```
Boolean variables need a prefix, per coding style.

cdumez replied:
Same applies to `m_blocked`

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,73 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+    Ref contentFilterCallbackAggregator = ContentFilterCallbackAggregator::create(*this, WTFMove(completionHandler));
+
+    for (Ref platformContentFilter : m_contentFilters) {
+        CompletionHandler<void(ResourceRequest, PlatformContentFilter::State)> completion = [contentFilterCallbackAggregator, platformContentFilter] (ResourceRequest&& request, PlatformContentFilter::State state) mutable {
```
A completion handler always needs to get called or we crash. Yet, you don't seem to call the completion handler when `!platformContentFilter->needsMoreData()`?

pvollan replied:
Ah, good catch, will fix!

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,73 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+
+        if (platformContentFilter->needsMoreData()) {
+            auto requestCopy = request;
+            platformContentFilter->willSendRequest(WTFMove(requestCopy), redirectResponse, WTFMove(completion));
```
Maybe we can use `ResourceRequest { request }` instead of a local variable.

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,73 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+            auto requestCopy = request;
+            platformContentFilter->willSendRequest(WTFMove(requestCopy), redirectResponse, WTFMove(completion));
+        }
+    };
```
Why the semicolon here? Maybe I'm misreading the diff but it look like the end of the `for` loop.

cdumez commented on Source/WebCore/loader/ContentFilter.h:
```diff
@@ -89,6 +93,24 @@ class ContentFilter : public CanMakeWeakPtr<ContentFilter>, public CanMakeChecke
 private:
     using State = PlatformContentFilter::State;
 
+    class ContentFilterCallbackAggregator : public RefCounted<ContentFilterCallbackAggregator> {
```
Are we confident that we don't need `ThreadSafeRefCounted` here? I did not look enough in the code but I know there is a background thread involved so I want to make sure this never gets ref'd / deref'd on different threads.

pvollan replied:
`ThreadSafeRefCounted` should not be needed here, since `ContentFilterCallbackAggregator` is being captured in a completion handler on the main thread, and then the completion handler is moved to a background thread, before being moved back to the main thread again, and then finally called. However, to guard against future changes in this area, I think we can make it thread safe now.

cdumez commented on Source/WebCore/loader/ContentFilter.h:
```diff
@@ -89,6 +93,24 @@ class ContentFilter : public CanMakeWeakPtr<ContentFilter>, public CanMakeChecke
+    public:
+        static auto create(ContentFilter& contentFilter, CompletionHandler<void(ResourceRequest&&)>&& callback) { return adoptRef(*new ContentFilterCallbackAggregator(contentFilter, WTFMove(callback))); }
+
+        virtual ~ContentFilterCallbackAggregator();
```
Why is the destructor virtual??

pvollan replied:
Good point, this is not required, will fix!

cdumez commented on Source/WebCore/loader/ContentFilter.h:
```diff
@@ -89,6 +93,24 @@ class ContentFilter : public CanMakeWeakPtr<ContentFilter>, public CanMakeChecke
+
+        CheckedPtr<ContentFilter> m_contentFilter;
+        CompletionHandler<void(ResourceRequest&&)> m_callback;
+        unsigned m_platformFiltersAllowed { 0 };
```
I think the name is a little unclear for a `count`. Maybe something like `m_numberOfFiltersAllowed` or `m_filtersAllowedCount`

cdumez commented on Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
```diff
@@ -107,13 +110,62 @@
+    RetainPtr nsRequest = request.protectedNSURLRequest(DoNotUpdateHTTPBody);
+
+    auto blockPtr = makeBlockPtr([protectedThis = Ref { *this }, request = WTFMove(request), completionHandler = WTFMove(completionHandler)](NEFilterSourceStatus status, NSDictionary *decisionInfo) mutable {
+        RetainPtr<NSString> modifiedRequestURLString;
```
I assume this gets called on a background thread? If so, I would add a threading assertion (or code comment as fallback) to document that this gets called or may get called on a background thread.

pvollan replied:
The `request` object is only being moved on the background thread in the current implementation. To guard against future changes, it is sensible to do cross thread copy here.

cdumez commented on Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
```diff
@@ -107,13 +110,62 @@
+
+            URL modifiedRequestURL { modifiedRequestURLString.get() };
+            if (!modifiedRequestURL.isValid()) {
+                LOG(ContentFiltering, "NetworkExtensionContentFilter failed to convert modified URL string %@ to a URL.\n", modifiedRequestURLString.get());
```
Can we do a RELEASE_LOG_ERROR here? It feels bad to silently ignore the URL in this case.

cdumez commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -2052,23 +2057,31 @@ void NetworkResourceLoader::startWithServiceWorker()
+        protectedThis->startRequest(WTFMove(request));
+    };
+
+    auto newRequest = ResourceRequest { originalRequest() };
```
`ResourceRequest newRequest { originalRequest() };`

cdumez reviewed (changes requested):
I made several comments, mostly minors but I do think the completion handler not getting called is a bug.

cdumez commented on Source/WebCore/loader/ContentFilter.h:
```diff
@@ -89,6 +93,24 @@ class ContentFilter : public CanMakeWeakPtr<ContentFilter>, public CanMakeChecke
+        void didReceivePlatformContentFilterDecision(PlatformContentFilter&, ResourceRequest&&, PlatformContentFilter::State);
+
+    private:
+        explicit ContentFilterCallbackAggregator(ContentFilter&, CompletionHandler<void(ResourceRequest&&)>&&);
```
nit: I don't think `explicit` is very useful here.

cdumez commented on Source/WebCore/loader/ContentFilter.h:
```diff
@@ -89,6 +93,24 @@ class ContentFilter : public CanMakeWeakPtr<ContentFilter>, public CanMakeChecke
+
+        CheckedPtr<ContentFilter> m_contentFilter;
+        CompletionHandler<void(ResourceRequest&&)> m_callback;
+        unsigned m_numberOfFiltersAllowed { 0 };
```
nit: Can we move this data member below `m_request` for likely better bit packing?

cdumez commented on Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
```diff
@@ -107,13 +110,65 @@
+
+    RetainPtr nsRequest = request.protectedNSURLRequest(DoNotUpdateHTTPBody);
+
+    auto blockPtr = makeBlockPtr([protectedThis = Ref { *this }, request = crossThreadCopy(WTFMove(request)), completionHandler = WTFMove(completionHandler)](NEFilterSourceStatus status, NSDictionary *decisionInfo) mutable {
```
We don't need crossThreadCopy() for `request`. I had removed this review comment after noticing it was a NSURLRequest instead of a ResourceRequest.

cdumez commented on Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
```diff
@@ -107,13 +110,65 @@
+        if (decisionInfo)
+            modifiedRequestURLString = checked_objc_cast<NSString>(decisionInfo[NEFilterSourceOptionsRedirectURL]);
+
+        RunLoop::mainSingleton().dispatch([status, decisionInfo = RetainPtr { decisionInfo }, modifiedRequestURLString = WTFMove(modifiedRequestURLString), protectedThis = WTFMove(protectedThis), request = crossThreadCopy(WTFMove(request)), completionHandler = WTFMove(completionHandler)] mutable {
```
Same comment about `crossThreadCopy()` not being needed.

cdumez commented on Source/WebCore/testing/MockContentFilter.cpp:
```diff
@@ -87,13 +93,22 @@ void MockContentFilter::willSendRequest(ResourceRequest& request, const Resource
 
+void MockContentFilter::willSendRequest(ResourceRequest&& request, const ResourceResponse& redirectResponse, CompletionHandler<void(ResourceRequest&&, State)>&& completionHandler)
+{
+    CompletionHandler<void()> completion = [&request, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler)] mutable {
```
This function seems very overkill with a CompletionHandler and a CompletionHandlerCallingScope, why can't we just write this?
```
willSendRequest(request, redirectResponse);
completionHandler(WTFMove(request), m_state);
```

cdumez reviewed (approved):
LGTM with fixes.

szewai commented on Source/WebCore/loader/ContentFilter.h:
```diff
@@ -89,6 +93,24 @@ class ContentFilter : public CanMakeWeakPtr<ContentFilter>, public CanMakeChecke
+    private:
+        explicit ContentFilterCallbackAggregator(ContentFilter&, CompletionHandler<void(ResourceRequest&&)>&&);
+
+        CheckedPtr<ContentFilter> m_contentFilter;
```
Why can we assume `ContentFilter` is during the lifetime of `ContentFilterCallbackAggregator`? (it seems `ContentFilter` does not own `ContentFilterCallbackAggregator`)

pvollan replied:
That's a good point! It is not obvious from the patch, but the `NetworkResourceLoader` is holding a unique pointer to the `ContentFilter`, and the completion handler is capturing a reference to the `NetworkResourceLoader` in `NetworkResourceLoader::startWithServiceWorker()`.

szewai commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,76 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+    ASSERT(RunLoop::isMain());
+
+    // Return early if no platform content filters are enabled.
+    if (!anyContentFilterIsEnabled()) {
```
Is this check necessary? If no content filter is enabled, shouldn't `PlatformContentFilter` directly invoke completionHandler with allowed?

pvollan replied:
That's a good point! I'll address this issue and the other comments shortly.

Thanks for reviewing!

szewai commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,76 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+
+    // Return early if no platform content filters are enabled.
+    if (!anyContentFilterIsEnabled()) {
+        continueAfterWillSendRequest(request, redirectResponse);
```
Why do we need to invoke the sync function again is no content filter is enabled?

pvollan replied:
This has been removed in the latest patch.

szewai commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,76 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+        return;
+    }
+
+    LOG(ContentFiltering, "ContentFilter received request for <%{sensitive}s> with redirect response from <%{sensitive}s>.\n", request.url().string().utf8().data(), redirectResponse.url().string().utf8().data());
```
`SENSITIVE_LOG_STRING` to replace `%{sensitive}`

szewai commented on Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
```diff
@@ -107,13 +110,65 @@
+            modifiedRequestURLString = checked_objc_cast<NSString>(decisionInfo[NEFilterSourceOptionsRedirectURL]);
+
+        RunLoop::mainSingleton().dispatch([status, decisionInfo = RetainPtr { decisionInfo }, modifiedRequestURLString = WTFMove(modifiedRequestURLString), protectedThis = WTFMove(protectedThis), request = crossThreadCopy(WTFMove(request)), completionHandler = WTFMove(completionHandler)] mutable {
+            protectedThis->handleDecision(status, replacementDataFromDecisionInfo(decisionInfo.get()));
```
So `handleDecision` can be invoked on both main thread and background thread now but we don't have lock around data members; is it safe?

pvollan replied:
Good catch! I am moving this back to the background thread in the upcoming patch.

szewai commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,76 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+        };
+
+        ASSERT(platformContentFilter->needsMoreData());
+        platformContentFilter->willSendRequest(ResourceRequest { request }, redirectResponse, WTFMove(completion));
```
In the existing implementation, I think we use updated request to query the next `PlatformContentFilter`; but in the new implementation we ask all `PlatformContentFilter`s with the original request. I am not sure if this is expected.

pvollan replied:
Good point! IMHO, I think that change makes sense, since the request passed to the platform content filter does not depend on the ordering of the platform filters, then.

szewai commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,76 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+    m_callback(WTFMove(m_request));
+}
+
+void ContentFilter::ContentFilterCallbackAggregator::didReceivePlatformContentFilterDecision(PlatformContentFilter& platformContentFilter, ResourceRequest&& request, PlatformContentFilter::State state)
```
Do we need both `PlatformContentFilter` and state as parameters? Can we just ask `PlatformContentFilter` for state?

szewai commented on Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
```diff
@@ -107,13 +110,65 @@
+            }
+
+            request.setURL(WTFMove(modifiedRequestURL));
+            completionHandler(WTFMove(request), protectedThis->m_state);
```
If the content filter will just update URL can we just pass url string back to the callback aggregator instead of copying the whole request?

pvollan replied:
That's a good point! I have updated the implementation.

szewai commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,71 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+    Ref contentFilterCallbackAggregator = ContentFilterCallbackAggregator::create(*this, request, WTFMove(completionHandler));
+
+    for (Ref platformContentFilter : m_contentFilters) {
+        CompletionHandler<void(String&&, PlatformContentFilter::State)> completion = [contentFilterCallbackAggregator, platformContentFilter] (String&& urlString, PlatformContentFilter::State) mutable {
```
`PlatformContentFilter::State` is unused now and should be removed

pvollan replied:
I will fix in the upcoming patch.

Thanks for reviewing!

cdumez commented on Source/WebCore/loader/ContentFilter.cpp:
```diff
@@ -123,6 +123,73 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
+    Ref contentFilterCallbackAggregator = ContentFilterCallbackAggregator::create(*this, request, WTFMove(completionHandler));
+
+    for (Ref platformContentFilter : m_contentFilters) {
+        CompletionHandler<void(String&&)> completion = [contentFilterCallbackAggregator, platformContentFilter] (String&& urlString) mutable {
```
No space between `] (`

cdumez reviewed (approved):
Still LGTM

pvollan committed:
The ContentFilter method continueAfterWillSendRequest should use a completion handler
https://bugs.webkit.org/show_bug.cgi?id=301189
rdar://151461666

Reviewed by Chris Dumez and Sihui Liu.

The ContentFilter method continueAfterWillSendRequest should use a completion handler to avoid blocking
the main thread waiting for a decision that can take a significant amount of time. In particular, it is
the NetworkExtensionContentFilter that can take a long time. For this reason, it is currently doing the
work on a non-main thread, and we have a binary semaphore that we use to wait synchronously for the
decision. Performing this work on the main thread can block other network requests. To avoid blocking
the main thread waiting for a semaphore, we should instead use a completion handler. This patch adds no
new threads, but is instead introducing a completion handler that is called when the decision is ready.

Before this change, the responsiveness timer in the UI process would terminate the Networking process
if the content filter was blocking the main thread for more than 6s. This issue is resolved with this
patch.

Test: contentfiltering/delay-willsendrequest-decision.html

* Source/WebCore/loader/ContentFilter.cpp:
(WebCore::ContentFilter::continueAfterWillSendRequest):
(WebCore::ContentFilter::ContentFilterCallbackAggregator::~ContentFilterCallbackAggregator):
(WebCore::ContentFilter::ContentFilterCallbackAggregator::didReceivePlatformContentFilterDecision):
(WebCore::ContentFilter::ContentFilterCallbackAggregator::ContentFilterCallbackAggregator):
(WebCore::ContentFilter::anyContentFilterIsEnabled const):
* Source/WebCore/loader/ContentFilter.h:
(WebCore::ContentFilter::ContentFilterCallbackAggregator::create):
* Source/WebCore/platform/PlatformContentFilter.h:
(WebCore::PlatformContentFilter::isEnabled const):
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.h:
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
(WebCore::NetworkExtensionContentFilter::willSendRequest):
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
(WebCore::ParentalControlsContentFilter::willSendRequest):
* Source/WebCore/testing/MockContentFilter.cpp:
(WebCore::MockContentFilter::willSendRequest):
* Source/WebCore/testing/MockContentFilter.h:
* Source/WebCore/testing/MockContentFilterSettings.cpp:
(WebCore::MockContentFilterSettings::setWillSendRequestDecisionDelay):
* Source/WebCore/testing/MockContentFilterSettings.h:
(WebCore::MockContentFilterSettings::willSendRequestDecisionDelay const):
* Source/WebCore/testing/MockContentFilterSettings.idl:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startContentFiltering):
(WebKit::NetworkResourceLoader::startWithServiceWorker):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/302627@main
